ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh

```diff
@@ -13,31 +13,82 @@ CONFIG="${SHARED_DIR}/install-config.yaml"
 
 expiration_date=$(date -d '8 hours' --iso=minutes --utc)
 
+declare -g MASTER_FAMILY
+declare -g MASTER_TYPE
+
 function join_by { local IFS="$1"; shift; echo "$*"; }
 
+function get_master_size {
+  master_size=null
+  if [[ "${SIZE_VARIANT}" == "xlarge" ]]; then
+    master_size=8xlarge
+  elif [[ "${SIZE_VARIANT}" == "large" ]]; then
+    master_size=4xlarge
+  elif [[ "${SIZE_VARIANT}" == "compact" ]]; then
+    master_size=2xlarge
+  fi
+  echo ${master_size}
+}
+
+function set_master_type {
+  local size
+  master_type=null
+  size=$(get_master_size)
+  if [[ "${size}" == "null" ]]; then
+    MASTER_TYPE="${size}"
+    return
+  fi
+  MASTER_TYPE="${MASTER_FAMILY}.$(get_master_size)"
+}
+
+function choose_preferred_type {
+  local type_preferred
+  local type_backup
+  local zones_cnt_pref
+  local zones_cnt_backup
+
+  type_preferred="${1}.$(get_master_size)"
+  type_backup="${2}.$(get_master_size)"
+
+  zones_cnt_pref=$(aws ec2 describe-instance-type-offerings --region "${REGION}" --location-type availability-zone --filters Name=instance-type,Values="${type_preferred}" --query 'InstanceTypeOfferings[].Location' --output json | jq -r '. | length')
+  zones_cnt_backup=$(aws ec2 describe-instance-type-offerings --region "${REGION}" --location-type availability-zone --filters Name=instance-type,Values="${type_backup}" --query 'InstanceTypeOfferings[].Location' --output json | jq -r '. | length')
+
+  if [[ ${zones_cnt_pref} -ge ${zones_cnt_backup} ]]; then
```

Comment:
In practice this probably ends up the same, but I would have expected that the preferred family be used as long as there was at least one zone (or perhaps 2 zones) offering the preferred family.

Reply:
iiuc, I am not sure if it will ends up the same, it will help to select the correct instance on the region which does not support the preferred one (see the [original PR](https://github.com/openshift/installer/pull/5327) for the table of availability):

output of `$ cat install-config.yaml |yq -r '. | (.platform.aws, .controlPlane.platform.aws)'`,

for `LEASED_RESOURCE=sa-east-1`:
``` json
{
  "region": "sa-east-1",
  "userTags": {
    "expirationDate": "2021-12-01T22:34+00:00"
  }
}
{
  "zones": [],
  "type": "m5.2xlarge"
}

```

for `LEASED_RESOURCE=us-east-1`:
```json
{
  "region": "us-east-1",
  "userTags": {
    "expirationDate": "2021-12-01T22:35+00:00"
  }
}
{
  "zones": [],
  "type": "m6i.2xlarge"
}
```

Reply:
> but I would have expected that the preferred family be used as long as there was at least one zone (or perhaps 2 zones)

if it's not a problem, we could change to something like:

``` shell
if [[ ${zones_cnt_pref} -ge 1 ]]; then
```

wdyt?

Reply:
updating the latest availability:
```
                count(m5.xlarge)  count(m6i.xlarge)  diff(m5_m6i)
region                                                           
af-south-1                     3                  0             3
ap-east-1                      3                  0             3
ap-northeast-1                 3                  2             1
ap-northeast-2                 4                  3             1
ap-northeast-3                 3                  0             3
ap-south-1                     3                  3             0
ap-southeast-1                 3                  3             0
ap-southeast-2                 3                  2             1
ca-central-1                   3                  0             3
eu-central-1                   3                  3             0
eu-north-1                     3                  0             3
eu-south-1                     3                  0             3
eu-west-1                      3                  3             0
eu-west-2                      3                  0             3
eu-west-3                      3                  3             0
me-south-1                     3                  0             3
sa-east-1                      3                  2             1
us-east-1                      5                  5             0
us-east-2                      3                  3             0
us-west-1                      2                  2             0
us-west-2                      6                  4             2

```